### PR TITLE
DO NOT MERGE - WIP: Add ability to serialize/deserialze astroid.Module (and friends)

### PR DESCRIPTION
**Pull request by thejcannon**
## Steps

- [ ] For new features or bug fixes, add a ChangeLog entry describing what your PR does.
- [ ] Write a good description on what the PR does.

## Description

This PR adds the ability to serialize and deserialize `astroid.Module` instances (and by association, every other kind of instance it may reference) to-and-from JSON. 

### Why persistence?

Persisting processed modules is useful primarily when parsing a large number of files, or files with a lot of interdependencies. So long as desrializing form the cache is quicker than parsing the files, it's a net win.

(I'll do some profiling with `pylint` on our codebase of about 1k files when this is more polished) 

### Why JSON, why not YAML, XML, or `pickle`?

JSON was chosen as it is compact (as opposed to yml or xml or similar) and we can choose what and how to store information. Although pickling might be simpler in terms of complexity, it is inherently insecure as a bad actor only needs to infect the `astroid` pickled data on disk to infect the next parse. Using JSON, since we control the data, the actor would have to infect the cache, and the Python environment the next parse runs in.

And since we aren't inventing anything new here, it should also be stated `mypy`'s cache is also JSON :wink:

### Changes

The bulk of the relevant changes have been localized to the `_persistence` module, which has been specially implemented in order to reduce the amount of changes to all of the relevant `astroid` classes. Additional changes have been made in order to either serialize/deserialize in a special way or to enable the default way.

A future change will have the changes to the module cache to leverage this new feature.

### Testing

TODO

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
|    | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Related Issue

Related #1145

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+from functools import wraps
+
+import astroid.builder
+import astroid.nodes
+
+def pytest_addoption(parser):
+    parser.addoption(
+        "--test-roundtrip-persistence", type=bool, default=True,
```
Comment (thejcannon) [inline tests/conftest.py]: Eventually, I think we'll add a tox section to re-run all tests with this enabled. For now it's just enabled so I can test easily.

**Review by thejcannon — COMMENTED** [review]
Not all tests pass right now. Latest errors are from inference not working (and since inference is quite the beasty, I thought I'd publish this PR as-is for initial feedback)

**Comment by thejcannon** [comment]
I'll use this PR to update progress and dump thoughts as I work through the remaining problems.

The current test is failing AFAICT due to `_explicit_inference` not persisting after serialization. I assume this should come from the 
`inference_tip` cache? :thinking:

File: astroid/nodes/scoped_nodes.py [hunk 2]
```diff
@@ -2096,8 +2155,7 @@ def postinit(
         :param keywords: The keywords given to the class definition.
         :type keywords: list(Keyword) or None
         """
-        if keywords is not None:
-            self.keywords = keywords
+        self.keywords = keywords
```
Comment (Pierre-Sassoulas) [inline astroid/nodes/scoped_nodes.py:2158]: I wonder why it was that way at first. Thanks you for the cleanup.

File: astroid/nodes/scoped_nodes.py [hunk 3]
```diff
@@ -3045,3 +3103,12 @@ def _get_assign_nodes(self):
             child_node._get_assign_nodes() for child_node in self.body
         )
         return list(itertools.chain.from_iterable(children_assign_nodes))
+
+    def __dump__(self, dumper):
+        data = super().__dump__(dumper)
```
Comment (Pierre-Sassoulas) [inline astroid/nodes/scoped_nodes.py:3108]: Is this a cultural reference to dumb and dumber ;) ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you for opening this. Sorry I don't have time for an in depth review right now. I agree with the need to improve the caching though. Could you be more specific about the advantages of JSON compared to pickle and about what make the persistance necessary ? Is it for very big code base or low ram machine where the cache would be bigger than ram ? Also could you add typing on newly added functions, please ?

File: astroid/nodes/scoped_nodes.py (hunk 3 above)
Comment (thejcannon) [reply astroid/nodes/scoped_nodes.py:3108]: Haha no, but we can pretend it is.
I originally was passing the remap around but thought that was kind of dirty, so wanted to pass a function that did the dumping and remembered the remap so "dumper" was the first word that came to mind 😂

**Comment by thejcannon** [comment]
I agree the AST parsing caching itself wouldn't be sufficient (hence why there isn't any code that actually persists the AST to disk yet). I assumed this would be the first step that would eventually lead to inference caching (since the inference results would need to be persistable).

Closing, though as I won't have much free time anymore to walk this along. It can be used as a starting point for that future effort though :tada:

**Pull request closed without merging**